httpbp/client_middlewares.go

```diff
@@ -0,0 +1,228 @@
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(roundTripper http.RoundTripper) http.RoundTripper
+
+type roundTripper func(req *http.Request) (*http.Response, error)
+
+func (f roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config Config, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		CircuitBreaker(config),
+	}
+
+	middlewares = append(middlewares, defaults...)
+
+	return &http.Client{
+		Transport: applyMiddlewares(transport, middlewares...),
+	}
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config Config) ClientMiddleware {
+	return func(roundTripper http.RoundTripper) http.RoundTripper {
+		return circuitBreaker.RoundTrip(roundTripper.RoundTrip, config.CircuitBreaker)
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+type circuitBreaker func(req *http.Request) (*http.Response, error)
+
+func (f circuitBreaker) RoundTrip(config CircuitBreakerConfig) roundTripper {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return &gobreaker.Settings{
+				Interval: 60 * time.Second,
+				ReadyToTrip: func(counts gobreaker.Counts) bool {
+					failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+					return counts.Requests >= config.MinRequests && failureRatio >= config.MinFailureRatio
+				},
+			}
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(req *http.Request) (*http.Response, error) {
+		host := req.URL.Hostname()
+
+		// new circuit breakers should rarely get allocated: the pool reduces
+		// garbage collector overhead
+		newBreaker := pool.Get()
+		breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+		if loaded {
+			pool.Put(newBreaker)
+		}
+
+		resp, err := breaker.(*gobreaker.CircuitBreaker).Execute(func() (interface{}, error) {
+			resp, err := f(req)
+			if err != nil {
+				return nil, err
+			}
+			// circuit break on any HTTP 5xx code
+			if resp.StatusCode >= http.StatusInternalServerError {
+				err = resp.Body.Close()
+				if err != nil {
+					return nil, err
+				}
+				return nil, err
```

Comment:
believe I asked this question before in the other PR as well, but not sure if it has been answered. Wouldn't the `err` here be `nil` since the body closure was fine, but we still it to have a error since we consider all 5xx to potentially break the circuit?

Reply:
If we don't add a middleware with [`httpbp.ClientErrorFromResponse`](https://pkg.go.dev/github.com/reddit/baseplate.go/httpbp#ClientErrorFromResponse) after this one, then yes go's stdlib http client would only return an error when there's network level error. But we should add such a middleware so this should still work (even without that, it still works for network errors).

Reply:
now since we can return both non-nil resp and err, this should be `return resp, err` instead.

Reply:
^ this comment is not addressed

Reply:
Apparently, round trippers are not supposed to pass a response if `err` is non-nil https://github.com/golang/go/blob/master/src/net/http/client.go#L256 :slightly_frowning_face: